File: tutor/src/main/java/com/sm/tutor/domain/dto/LectureDto.java
package com.sm.tutor.domain.dto;

import com.sm.tutor.domain.Lecture;
import com.sm.tutor.domain.LectureAge;
import com.sm.tutor.domain.LectureImage;
import com.sm.tutor.domain.LectureLocation;
import com.sm.tutor.domain.LectureReview;
import com.sm.tutor.domain.LectureTime;
import com.sm.tutor.domain.Tutor;
import java.util.List;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;

@Builder
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@ToString
public class LectureDto {

  private Integer id;
  private Integer tutorId;
  private Integer categoryId;
  private String title;
  private String content;
  private Boolean activation;
  private Integer online;
  private Integer tuitionMaximum;
  private Integer tuitionMinimum;
  private Integer tuteeNumber;
  private Integer gender;
  private Integer level;
  private List<LectureAgeDto> ages;
  private List<LectureImageDto> images;
  private List<LectureLocationDto> locations;
  private List<LectureReviewDto> reviews;
  private List<LectureTimeDto> times;

  public static LectureDto fromEntity(Lecture lecture) {
    if (lecture == null) {
      return null;
    }

    return LectureDto.builder()
        .id(lecture.getId())
        .tutorId(lecture.getTutor().getId())
        .categoryId(lecture.getCategoryId())
        .title(lecture.getTitle())
        .content(lecture.getContent())
        .activation(lecture.getActivation())
        .online(lecture.getOnline())
        .tuitionMaximum(lecture.getTuitionMaximum())
        .tuitionMinimum(lecture.getTuitionMinimum())
        .tuteeNumber(lecture.getTuteeNumber())
        .gender(lecture.getGender())
        .level(lecture.getLevel())
        .build();
  }

  public Lecture toEntity(Tutor tutor, List<LectureAge> ages, List<LectureImage> images,
      List<LectureLocation> locations, List<LectureReview> reviews, List<LectureTime> times) {
    return new Lecture(tutor, this.categoryId, this.title, this.content, this.activation,
        this.online, this.tuitionMaximum,
        this.tuitionMinimum, this.tuteeNumber, this.gender, this.level, ages, images, locations,
        reviews, times);
  }
}

File: frontend/lib/main/domain/model/dto/lecture_dto.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:tutor_platform/core/properties/lecture_information.dart';
import 'package:tutor_platform/main/domain/model/dto/lecture_age_dto.dart';
import 'package:tutor_platform/main/domain/model/dto/lecture_image_dto.dart';
import 'package:tutor_platform/main/domain/model/dto/lecture_location_dto.dart';
import 'package:tutor_platform/main/domain/model/dto/lecture_review_dto.dart';
import 'package:tutor_platform/main/domain/model/dto/lecture_time_dto.dart';

part 'lecture_dto.freezed.dart';

part 'lecture_dto.g.dart';

@freezed
class LectureDto with _$LectureDto {
  const LectureDto._();

  const factory LectureDto({
    required int id,
    required int tutorId,
    required int categoryId,
    required String title,
    required String content,
    required bool activation,
    required int online,
    required int tuitionMaximum,
    required int tuitionMinimum,
    required int tuteeNumber,
    required int? gender,
    required int level,
    required List<LectureAgeDto>? ages,
    required List<LectureImageDto>? images,
    @Default([]) List<LectureLocationDto>? locations,
    required List<LectureReviewDto>? reviews,
    required List<LectureTimeDto>? times,
  }) = _LectureDto;

  factory LectureDto.fromJson(Map<String, dynamic> json) =>
      _$LectureDtoFromJson(json);

  double? get avgRating => reviews!.isEmpty
      ? null
      : reviews!.map((e) => e.score).reduce((a, b) => a + b) / reviews!.length;

  String? get mainImage => images!.firstWhere((e) => e.mainImage, orElse: () => images!.isEmpty ? LectureImageDto.empty() : images![0]).image;

  List<String> get subImages => images!.where((e) => !e.mainImage).map((e) => e.image).toList();

  OnOffline get onlineOrOffline => OnOffline.values[online];
}
